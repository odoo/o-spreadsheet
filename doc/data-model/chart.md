# ChartPlugin: Chart Data Structure

The ChartPlugin manages chart figures in o-spreadsheet. Charts are stored as entries in the `figures` array of each sheet object, with a `tag` of `"chart"` and a `data` object describing the chart configuration.

---

## Structure

- **Per-sheet `figures` array:**
  - Each entry is a figure object.
  - Chart figures have `tag: "chart"` and a `data` object.

The complete definition of each chart is complex, the details are defined in [src/types/chart/chart.ts](src/types/chart/chart.ts).
To have a detailed chart, create it in o-spreadsheet and export the data to see the full structure.

---

## Chart Figure Object

```js
{
  "id": string,           // Unique figure ID
  "tag": "chart",         // Identifies this figure as a chart

  // Chart site and positionning
  "col": number,          // Column of the anchor of the chart
  "row": number,          // Row of the anchor of the chart
  "offset" {              // Position starting from the col andd row in pixels
    "x": number,
    "y": number,
  },
  "width": number,        // Chart width in pixels
  "height": number,       // Chart height in pixels

  // Definition of the chart
  "data": {
    "type": string,       // Chart type (e.g., "line", "bar", "pie", "scatter", "combo", etc.)
    "dataSetsHaveTitle": boolean, // Whether datasets have titles
    "background": string, // Background color (hex code)
    "dataSetsHaveTitle": false, // (optional) Whether this dataset has a title
    "dataSets": [         // Array of dataset objects
      {
        "type": string,      // (optional) For combo charts: dataset type (e.g., "bar", "line")
        "dataRange": string, // Cell range for data (e.g., "Sheet1!B26:B35")
        "yAxisId": "y",
        "label": string,     // Label of the data series
        "backgroundColor": string,
        "trend": {
            "type": "polynomial" | "exponential" | "logarithmic" | "trailingMovingAverage",
            "display": true,
            "order": number, // (optional)
            "color": string, // (optional)
            "window": number // (optional)
        }
      },
      // ...
    ],
    // Cell range for labels
    "legendPosition": "top"|"left"|"bottom"|"right"|"none", "labelRange": string,
    "title": {                      // Chart title
        "text": string,
        "bold": boolean,
        "italic": boolean,
        "align": "left" | "center" | "right",        // Text alignment
        "verticalAlign": "top" | "middle" | "bottom",// Vertical alignment
        "fontSize": number,          // Font size in pixels
        "color": string,             // Text color (hex code)
        "fillColor": string,         // Background color for the title
     },
    "stacked": boolean (for bar/line/combo),
    "aggregated": boolean,
    "humanizeLargeNumbers": boolean, // Whether to humanize large numbers (e.g., 1K, 1M)

    // Additional chart-type-specific options:
    // - "region": string (for geo charts)
    // - "verticalAxisPosition": string (for waterfall)
    // - "showSubTotals": boolean
    // - "showConnectorLines": boolean
    // - etc.
  }
}
```

---

## Example

```js
"figures": [
    {
        "id": "c85dcad9-ca99",
        "col": 2,
        "row": 11,
        "offset": {
            "x": 52.5,
            "y": 3
        },
        "width": 536,
        "height": 335,
        "tag": "chart",
        "data": {
        "type": "bar",
        "dataSetsHaveTitle": true,
        "dataSets": [
            {
                "dataRange": "B2:B16",
                "yAxisId": "y"
            }
        ],
        "legendPosition": "none",
        "labelRange": "A2:A16",
        "title": {},
        "stacked": false,
        "aggregated": false
        }
    }
]
```

---

## Notes

- Chart figures are always stored in the `figures` array of a sheet, not at the top level.
- The `data` object can include additional properties depending on the chart type (see examples in demo/data.js).
- Chart IDs must be unique within the sheet.
- The `x` and `y` properties control the chart's position on the canvas (optional).
- For more advanced chart types (e.g., combo, geo, waterfall), see the full data structure in demo/data.js.

---

## Carousel Data Model

A carousel is a figure that can contain multiple charts (and optionally data views) and allows users to switch between them. Carousel figures are stored in the `figures` array of a sheet object, with a `tag` of `"carousel"` and a `data` object describing the carousel configuration.

### Carousel Figure Object

```js
{
  "id": string,           // Unique figure ID
  "tag": "carousel",      // Identifies this figure as a carousel
  "col": number,          // Column of the anchor of the carousel
  "row": number,          // Row of the anchor of the carousel
  "offset": {             // Position starting from the col and row in pixels
    "x": number,
    "y": number
  },
  "width": number,        // Carousel width in pixels
  "height": number,       // Carousel height in pixels
  "data": {
    "chartDefinitions": { // (optional) Map of chartId to chart definition
      [chartId: string]: { ...chartDefinition }
    },
    "items": [            // Array of carousel items
      {
        "type": "chart",      // Item type: chart
        "chartId": string,    // ID referencing a chart definition in chartDefinitions
        "carouselTitle": {    // (optional) Title design for this chart in the carousel
          "text": string,
          "bold": boolean,
          "italic": boolean,
          "align": "left"|"center"|"right",
          "verticalAlign": "top"|"middle"|"bottom",
          "fontSize": number,
          "color": string,
          "fillColor": string
        }
      },
      {
        "type": "carouselDataView" // Item type: data view (optional)
      }
      // ...
    ]
  }
}
```

#### Notes on `chartDefinitions`

- The `chartDefinitions` object maps chart IDs to their chart configuration.
- Each carousel item of type `"chart"` references its chart definition by `chartId`.
- This allows the carousel to manage multiple charts internally, without requiring separate chart figures.

### Example

```js
"figures": [
  {
    "id": "carousel-1",
    "col": 1,
    "row": 5,
    "offset": { "x": 0, "y": 0 },
    "width": 600,
    "height": 400,
    "tag": "carousel",
    "data": {
      "chartDefinitions": {
        "1": {
          "type": "line",
          "dataSetsHaveTitle": true,
          "background": "#FFFFFF",
          "dataSets": [
            { "dataRange": "Sheet1!B26:B35" },
            { "dataRange": "Sheet1!C26:C35" }
          ],
          "legendPosition": "top",
          "labelRange": "Sheet1!A27:A35",
          "title": {},
          "stacked": false,
          "humanize": true
        },
        "2": {
          "type": "bar",
          "dataSetsHaveTitle": false,
          "background": "#FFFFFF",
          "dataSets": [
            { "dataRange": "Sheet1!B27:B35" },
            { "dataRange": "Sheet1!C27:C35" }
          ],
          "legendPosition": "top",
          "labelRange": "Sheet1!A27:A35",
          "title": {},
          "stacked": false,
          "humanize": true
        }
      },
      "items": [
        {
          "type": "chart",
          "chartId": "1",
          "carouselTitle": { "text": "Line" }
        },
        {
          "type": "chart",
          "chartId": "2",
          "carouselTitle": { "text": "Bar" }
        }
      ]
    }
  }
]
```

### Notes

- Carousel figures are stored in the `figures` array of a sheet, similar to chart figures.
- The `items` array can contain multiple charts and/or data views.
- Each chart item references a chart by its `chartId`, which must exist in `chartDefinitions`.
- The optional `carouselTitle` allows customizing the title for each chart within the carousel.
- Carousel IDs must be unique within the sheet.
- See the full structure in demo/data.js for advanced usage.
