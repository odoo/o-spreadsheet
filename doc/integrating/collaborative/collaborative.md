# Collaborative Edition

Realtime collaboration edition can be enabled to synchronize a spreadsheet across multiple clients.
It is enabled by providing a way to communicate with other connected clients to the `<Spreadsheet/>` component. We call it the _transport service_. Its interface and how it should be implemented is described in a [dedicated section](#transport-service). An optional [client](tsdoc/interfaces/client.md) can also be provided to display the names of connected clients.

```ts
const model = new Model(data, {
  transportService,
  client: { id: 456, name: "Raoul" },
});
```

```xml
<Spreadsheet model="model"/>
```

To learn more about the explored possibilities, please read [this](collaborative_choices.md).

- [Collaborative Edition](#collaborative-edition)
  - [Transport Service](#transport-service)
  - [Coordinating server](#coordinating-server)
  - [How it works](#how-it-works)
    - [Operational Transform](#operational-transform)
  - [Extension](#extension)
    - [Commands](#commands)
    - [Transformations](#transformations)
    - [History](#history)

## Transport Service

To enable realtime collaboration edition, a TransportService should be given to the Spreadsheet component.

The TransportService is responsible for sending and receiving messages between the clients with the help of central server.

The transport service should implement the [TransportService](tsdoc/interfaces/transportservice.md) interface.

An example can be found in demo/transport.js. The example is implemented using Websockets.

## Coordinating server

[Messages](tsdoc/README.md#CollaborationMessage) sent through the transport service should be handled by a centralized server. Its goal it to coordinate messages, order them and maintain the current spreadsheet revision.
For each received messages, the server decides if it's accepted or not. If the message is accepted, it's transmitted to all clients
(including the one which sent the message).

Messages can be split in two categories:

1. The messages that do not change the state of the spreadsheet and are always accepted:

- [ClientJoinedMessage](tsdoc/interfaces/clientjoinedmessage.md)
- [ClientMovedMessage](tsdoc/interfaces/clientmovedmessage.md)
- [ClientLeftMessage](tsdoc/interfaces/clientleftmessage.md)

1. The messages that require special care to ensure a correct message ordering:

- [RemoteRevisionMessage](tsdoc/interfaces/remoterevisionmessage.md)
- [RevisionUndoneMessage](tsdoc/interfaces/revisionundonemessage.md)
- [RevisionRedoneMessage](tsdoc/interfaces/revisionredonemessage.md)

They have two important properties: `serverRevisionId` and `nextRevisionId`. A message is accepted only if its `serverRevisionId`
is the same as the server revision id. The message is then transmitted to all clients, and the server revision id becomes `nextRevisionId`. Otherwise, the message is rejected. It means that the client should re-sent it later, after transforming it with all the messages it received in the meantime.

By convention, the initial revision id of a blank spreadsheet is `"START_REVISION"`.

A basic example can be found in tools/server/main.js. The example is implemented using Websockets

## How it works

Realtime collaborative editing is based on [Operational Transform](https://en.wikipedia.org/wiki/Operational_transformation).

### Operational Transform

The aim of operation transform is to manage realtime collaborative editing on a document and preserve consistency even with concurrent operations.
It ensures all clients eventually converge to the same state while preserving user intentions.

Let's take a look at two concurrent operations, why it can lead to problems and how operational transform can help:

Alice and Bob are working on the same sheet. At the same time, Alice wants to add a column before "B" and Bob wants to update the cell "B1" with the content "Hello". Let's call Alice's operation `OA` and Bob's operation `OB`.
Both Alice and Bob send their operation to the other.

When Alice receives `OB`, it would update "B1" with "Hello". When Bob receives `OA`,
it would add a column before "B", which moves the current column B to column C.
We now have:

- Alice with an additional column before "B" and "Hello" in "B1"
- Bob with an additional column before "B" and "Hello" in "C1"

Both state have diverged. This can't be good...

That's when operational transform can help.

`OB` on Alice's side is no longer valid since column B was moved to column C. The operation should be adapted to
reflect that. To do that, we apply a transformation to `OB`, to target the column C instead of column B. `OB` becomes "Update "C1" with the content `Hello`". With this transformation, Bob's intention is preserved and the states converge.

In order to identify and resolve which operations conflict, we need a way to detect two concurrent actions, and a way to order them.

To detect concurrency, we introduce a revision log, which is a unique identifier to indicate on which state the action is executed.
Each time an action is received and accepted by the server (the revision log of the server is the same as the action), the revision log is incremented.

Now, we need a way to transform the commands. An easy way would be to transform commands on the server, like [Google Drive](https://drive.googleblog.com/2010/09/whats-different-about-new-google-docs.html). But, in the context of using o-spreadsheet with Odoo, we didn't want to duplicate the transformations functions in both client and server, as the two are implemented in different language.

When a client executes an action, it's locally executed, sent to the server and kept in the pending actions of the user.
If the action is accepted by the server, the action is removed from the pending ones. In the other case, the pending actions are reverted, transformed with the action received from the server, re-applied locally and resent to the server until they are accepted.

Local undo is implemented by keeping locally the actions of the user, revert to just before the undo-ed action, transform the next actions as if the undo-ed action was not executed and re-apply them. Local redo follow the same logic.

To keep a synchronous state, we only need to share the commands which impacts the state of spreadsheet (columns, grid, ...) but not the local state (selection of the user, composer state, ...)

This solution has a lot of pros, but also some cons:

1. We need to write a transformation function for each command we create, which could theoretically becomes huge. However, in practice, the transformations only concerns commands which changes the grid (add/remove columns, remove a sheet, merge).
1. Undo/Redo is synchronous, i.e. it should be accepted by the server before being executed locally.

## Extension

To extend o-spreadsheet without breaking realtime collaborative edition, there are some points to keep in mind.

### Commands

There are two types of [commands](add_command.md): `CoreCommands` and `Commands`. Only `CoreCommands` are synchronized.
Here is the way to register a `CoreCommand` to o-spreadsheet.

```js
const { coreTypes } = o_spreadsheet;

coreTypes.add("MY_COMMAND_NAME");
```

### Transformations

For each new `CoreCommand`, a transformation function could be required for each other `CoreCommands`.

A transformation function takes as arguments the already executed command, and the command to transform. It should return a `CoreCommands` or `undefined` if the command should be skipped.

If a transformation is required, here is the way to declare it. The transformation check that the command is on the same sheet of the deleted sheet. If true, the command should be skipped.

```js
const { otRegistry } = o_spreadsheet.registries;

otRegistry.addTransformation("DELETE_SHEET", ["MY_COMMAND"], (myCommand, deleteSheet) => {
  // MY_COMMAND should takes DELETE_SHEET into account as DELETE_SHEET is arrived first
  if (myCommand.sheetId === deleteSheet.sheetId) {
    return undefined;
  }
  return myCommand;
});
```

### History

In addition to a transformation, an "inverse function" is required for commands which transforms other commands. The inverse function takes as input a `CoreCommand` and should return a list of `CoreCommands`. Those commands should have the inverse behavior as the original commands when used in transformation.

```txt
Given:
  A = a command
  B = another command
  A' = A transformed with B
  B⁻¹ = inverse of B
The inverse function should yield a command B⁻¹ such that:
  A === (A' transformed with B⁻¹)
```

In other words, transforming with `B` then transforming the result with `B⁻¹` should have no effect and yield the original command.

The inverse function is used during a selective undo to transform commands executed after the undo-ed command as if the command had never been executed.

Here is the way to declare it.

```js
const { inverseCommandRegistry } = o_spreadsheet.registries;

inverseCommandRegistry.add("CREATE_SHEET", (cmd) => {
  return [{ type: "DELETE_SHEET", sheetId: cmd.sheetId }];
});
```
