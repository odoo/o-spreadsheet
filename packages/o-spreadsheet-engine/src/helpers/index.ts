export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
