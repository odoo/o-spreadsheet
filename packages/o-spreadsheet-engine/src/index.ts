export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
