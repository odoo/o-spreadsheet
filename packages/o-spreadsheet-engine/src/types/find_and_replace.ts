import { Range } from "./range";

export interface SearchOptions {
  matchCase: boolean;
  exactMatch: boolean;
  searchFormulas: boolean;
  searchScope: "allSheets" | "activeSheet" | "specificRange";
  specificRange?: Range;
}
