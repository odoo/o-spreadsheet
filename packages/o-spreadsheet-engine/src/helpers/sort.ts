import { CellValue, CellValueType, EvaluatedCell, SortDirection } from "../types";

type CellWithIndex = { index: number; type: CellValueType; value: any };
const SORT_TYPES: CellValueType[] = [
  CellValueType.number,
  CellValueType.error,
  CellValueType.text,
  CellValueType.boolean,
];

export function cellsSortingCriterion(sortingOrder: string) {
  const inverse = sortingOrder === "asc" ? 1 : -1;
  return (
    left: { type: CellValueType; value: CellValue },
    right: { type: CellValueType; value: CellValue }
  ) => {
    if (left.type === CellValueType.empty) {
      return right.type === CellValueType.empty ? 0 : 1;
    } else if (right.type === CellValueType.empty) {
      return -1;
    }
    let typeOrder = SORT_TYPES.indexOf(left.type) - SORT_TYPES.indexOf(right.type);
    if (typeOrder === 0) {
      if (left.type === CellValueType.text || left.type === CellValueType.error) {
        typeOrder = (left.value as string).localeCompare(right.value as string);
      } else {
        typeOrder = (left.value as number) - (right.value as number);
      }
    }
    return inverse * typeOrder;
  };
}

export function sort(
  cells: EvaluatedCell[],
  sortDirection: SortDirection,
  emptyCellAsZero: boolean
): CellWithIndex[] {
  const cellsWithIndex: CellWithIndex[] = cells.map((cell, index) => ({
    index,
    type: cell.type,
    value: cell.value,
  }));

  const cellsToSort = emptyCellAsZero
    ? cellsWithIndex.map((cell) =>
        cell.type === CellValueType.empty ? { ...cell, type: CellValueType.number, value: 0 } : cell
      )
    : cellsWithIndex;

  return cellsToSort.sort(cellsSortingCriterion(sortDirection));
}
