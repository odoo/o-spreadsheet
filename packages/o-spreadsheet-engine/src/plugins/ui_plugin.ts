import { Session } from "../collaborative/session";
import { StateObserver } from "../state_observer";
import { ClientPosition } from "../types/collaborative/session";
import { Command, CommandDispatcher, CommandResult, CoreCommand } from "../types/commands";
import { Currency } from "../types/currency";
import { Getters } from "../types/getters";
import { HistoryChange } from "../types/history2";
import { Color } from "../types/misc";
import { ModelConfig } from "../types/model";
import { GridRenderingContext, LayerName } from "../types/rendering";
import { SelectionStreamProcessor } from "../types/selection_stream_processor";
import { ExcelWorkbookData } from "../types/workbook_data";
import { BasePlugin } from "./base_plugin";

export type UIActions = Pick<ModelConfig, "notifyUI" | "raiseBlockingErrorUI">;

export interface UIPluginConfig {
  readonly getters: Getters;
  readonly stateObserver: StateObserver<CoreCommand, HistoryChange>;
  readonly dispatch: CommandDispatcher["dispatch"];
  readonly canDispatch: CommandDispatcher["dispatch"];
  readonly selection: SelectionStreamProcessor;
  readonly moveClient: (position: ClientPosition) => void;
  readonly uiActions: UIActions;
  readonly custom: ModelConfig["custom"];
  readonly session: Session;
  readonly defaultCurrency?: Partial<Currency>;
  readonly customColors: Color[];
  readonly external: ModelConfig["external"];
}

export interface UIPluginConstructor {
  new (config: UIPluginConfig): UIPlugin;
  layers: Readonly<LayerName[]>;
  getters: readonly string[];
}

/**
 * UI plugins handle any transient data required to display a spreadsheet.
 * They can draw on the grid canvas.
 */
export class UIPlugin<State = any> extends BasePlugin<
  State,
  Command,
  CommandResult,
  HistoryChange,
  ExcelWorkbookData
> {
  static layers: Readonly<LayerName[]> = [];

  protected getters: Getters;
  protected ui: UIActions;
  protected selection: SelectionStreamProcessor;
  protected dispatch: CommandDispatcher["dispatch"];
  protected canDispatch: CommandDispatcher["dispatch"];

  constructor({
    getters,
    stateObserver,
    dispatch,
    canDispatch,
    uiActions,
    selection,
  }: UIPluginConfig) {
    super(stateObserver, CommandResult.Success);
    this.getters = getters;
    this.ui = uiActions;
    this.selection = selection;
    this.dispatch = dispatch;
    this.canDispatch = canDispatch;
  }

  // ---------------------------------------------------------------------------
  // Grid rendering
  // ---------------------------------------------------------------------------

  drawLayer(ctx: GridRenderingContext, layer: LayerName) {}
}
