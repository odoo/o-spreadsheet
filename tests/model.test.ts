import { CommandResult, CorePlugin } from "../src";
import { toZone } from "../src/helpers";
import { LocalHistory } from "../src/history/local_history";
import { Mode, Model } from "../src/model";
import { BordersPlugin } from "../src/plugins/core/borders";
import { CellPlugin } from "../src/plugins/core/cell";
import { ChartPlugin } from "../src/plugins/core/chart";
import { ConditionalFormatPlugin } from "../src/plugins/core/conditional_format";
import { FigurePlugin } from "../src/plugins/core/figures";
import { MergePlugin } from "../src/plugins/core/merge";
import { RangeAdapter } from "../src/plugins/core/range";
import { SheetPlugin } from "../src/plugins/core/sheet";
import { corePluginRegistry, uiPluginRegistry } from "../src/plugins/index";
import { AutomaticSumPlugin } from "../src/plugins/ui/automatic_sum";
import { FindAndReplacePlugin } from "../src/plugins/ui/find_and_replace";
import { SortPlugin } from "../src/plugins/ui/sort";
import { SheetUIPlugin } from "../src/plugins/ui/ui_sheet";
import { UIPlugin } from "../src/plugins/ui_plugin";
import { Command, CoreCommand, coreTypes, DispatchResult } from "../src/types";
import { setupCollaborativeEnv } from "./collaborative/collaborative_helpers";
import { selectCell, setCellContent } from "./test_helpers/commands_helpers";
import { getCell, getCellText } from "./test_helpers/getters_helpers";

function getNbrPlugin(mode: Mode): number {
  return (
    corePluginRegistry
      .getAll()
      .reduce((acc, plugin) => (plugin.modes.includes(mode) ? acc + 1 : acc), 0) +
    uiPluginRegistry
      .getAll()
      .reduce((acc, plugin) => (plugin.modes.includes(mode) ? acc + 1 : acc), 0)
  );
}

describe("Model", () => {
  test("can create model in headless mode", () => {
    const model = new Model({}, { mode: "headless" });
    expect(model["handlers"]).toHaveLength(13);
    expect(model["handlers"][0]).toBeInstanceOf(RangeAdapter);
    expect(model["handlers"][1]).toBeInstanceOf(SheetPlugin);
    expect(model["handlers"][2]).toBeInstanceOf(CellPlugin);
    expect(model["handlers"][3]).toBeInstanceOf(MergePlugin);
    expect(model["handlers"][4]).toBeInstanceOf(BordersPlugin);
    expect(model["handlers"][5]).toBeInstanceOf(ConditionalFormatPlugin);
    expect(model["handlers"][6]).toBeInstanceOf(FigurePlugin);
    expect(model["handlers"][7]).toBeInstanceOf(ChartPlugin);
    expect(model["handlers"][8]).toBeInstanceOf(SheetUIPlugin);
    expect(model["handlers"][9]).toBeInstanceOf(FindAndReplacePlugin);
    expect(model["handlers"][10]).toBeInstanceOf(SortPlugin);
    expect(model["handlers"][11]).toBeInstanceOf(AutomaticSumPlugin);
    expect(model["handlers"][12]).toBeInstanceOf(LocalHistory);
  });

  test("All plugin compatible with normal mode are loaded on normal mode", () => {
    const model = new Model();
    const nbr = getNbrPlugin("normal");
    expect(model["handlers"]).toHaveLength(nbr + 2); //+1 for Range +1 for History
  });

  test("All plugin compatible with headless mode are loaded on headless mode", () => {
    const model = new Model({}, { mode: "headless" });
    const nbr = getNbrPlugin("headless");
    expect(model["handlers"]).toHaveLength(nbr + 2); //+1 for Range +1 for History
  });

  test("Model in headless mode should not evaluate cells", () => {
    const model = new Model({ sheets: [{ id: "sheet1" }] }, { mode: "headless" });
    setCellContent(model, "A1", "=1", "sheet1");
    expect(getCell(model, "A1", "sheet1")!.evaluated.value).not.toBe("1");
  });

  test("can add a Plugin only in headless mode", () => {
    class NormalPlugin extends UIPlugin {
      static modes: Mode[] = ["normal"];
    }
    class HeadlessPlugin extends UIPlugin {
      static modes: Mode[] = ["headless"];
    }
    uiPluginRegistry.add("normalPlugin", NormalPlugin);
    uiPluginRegistry.add("headlessPlugin", HeadlessPlugin);
    const modelNormal = new Model();
    expect(modelNormal["handlers"].find((handler) => handler instanceof NormalPlugin)).toBeTruthy();
    expect(
      modelNormal["handlers"].find((handler) => handler instanceof HeadlessPlugin)
    ).toBeFalsy();
    const modelHeadless = new Model({}, { mode: "headless" });
    expect(
      modelHeadless["handlers"].find((handler) => handler instanceof NormalPlugin)
    ).toBeFalsy();
    expect(
      modelHeadless["handlers"].find((handler) => handler instanceof HeadlessPlugin)
    ).toBeTruthy();
  });

  test("core plugin can refuse command from UI plugin", () => {
    class MyCorePlugin extends CorePlugin {
      allowDispatch(cmd: CoreCommand) {
        if (cmd.type === "UPDATE_CELL") {
          return CommandResult.CancelledForUnknownReason;
        }
        return CommandResult.Success;
      }
    }
    let result: DispatchResult | undefined = undefined;
    class MyUIPlugin extends UIPlugin {
      handle(cmd: Command) {
        if (cmd.type === "COPY") {
          result = this.dispatch("UPDATE_CELL", {
            col: 0,
            row: 0,
            sheetId: this.getters.getActiveSheetId(),
            content: "hello",
          });
        }
      }
    }
    uiPluginRegistry.add("myUIPlugin", MyUIPlugin);
    corePluginRegistry.add("myCorePlugin", MyCorePlugin);
    const model = new Model();
    model.dispatch("COPY", { target: [toZone("A1")] });
    expect(result).toBeCancelledBecause(CommandResult.CancelledForUnknownReason);
    uiPluginRegistry.remove("myUIPlugin");
    corePluginRegistry.remove("myCorePlugin");
  });

  test("core plugin cannot refuse command from core plugin", () => {
    let result: DispatchResult | undefined = undefined;
    class MyCorePlugin extends CorePlugin {
      allowDispatch(cmd: CoreCommand) {
        if (cmd.type === "UPDATE_CELL") {
          return CommandResult.CancelledForUnknownReason;
        }
        return CommandResult.Success;
      }
      handle(cmd: CoreCommand) {
        if (cmd.type === "CREATE_SHEET") {
          result = this.dispatch("UPDATE_CELL", {
            col: 0,
            row: 0,
            sheetId: cmd.sheetId,
            content: "Hello",
          });
        }
      }
    }
    corePluginRegistry.add("myCorePlugin", MyCorePlugin);
    const model = new Model();
    model.dispatch("CREATE_SHEET", { sheetId: "42", position: 1 });
    expect(result).toBeSuccessfullyDispatched();
    expect(getCellText(model, "A1", "42")).toBe("Hello");
    corePluginRegistry.remove("myCorePlugin");
  });

  test("UI plugin cannot refuse command from UI plugin", () => {
    let result: DispatchResult | undefined = undefined;
    class MyUIPlugin extends UIPlugin {
      allowDispatch(cmd: Command) {
        if (cmd.type === "PASTE") {
          return CommandResult.CancelledForUnknownReason;
        }
        return CommandResult.Success;
      }
      handle(cmd: Command) {
        if (cmd.type === "COPY") {
          result = this.dispatch("PASTE", {
            target: [toZone("A2")],
          });
        }
      }
    }
    uiPluginRegistry.add("myUIPlugin", MyUIPlugin);
    const model = new Model();
    setCellContent(model, "A1", "copy&paste me");
    model.dispatch("COPY", { target: [toZone("A1")] });
    expect(result).toBeSuccessfullyDispatched();
    expect(getCellText(model, "A2")).toBe("copy&paste me");
    uiPluginRegistry.remove("myUIPlugin");
  });

  test("Can open a model in readonly mode", () => {
    const model = new Model({}, { isReadonly: true });
    expect(model.getters.isReadonly()).toBe(true);
  });

  test("Some commands are not dispatched in readonly mode", () => {
    const model = new Model({}, { isReadonly: true });
    expect(setCellContent(model, "A1", "hello")).toBeCancelledBecause(CommandResult.Readonly);
  });

  test("Moving the selection is allowed in readonly mode", () => {
    const model = new Model({}, { isReadonly: true });
    expect(selectCell(model, "A15")).toBeSuccessfullyDispatched();
  });

  test("Replayed commands are not send to UI plugins", () => {
    let numberCall = 0;
    //@ts-ignore
    coreTypes.add("MY_CMD_1");
    //@ts-ignore
    coreTypes.add("MY_CMD_2");
    class MyUIPlugin extends UIPlugin {
      handle(cmd: Command) {
        //@ts-ignore
        if (cmd.type === "MY_CMD_2") {
          if (this.getters.getClient().id === "bob") {
            numberCall++;
          }
        }
      }
    }
    uiPluginRegistry.add("myUIPlugin", MyUIPlugin);

    class MyCorePlugin extends CorePlugin {
      public readonly state: number = 0;
      handle(cmd: CoreCommand) {
        //@ts-ignore
        if (cmd.type === "MY_CMD_1") {
          this.history.update("state", 1);
          //@ts-ignore
          this.dispatch("MY_CMD_2");
        }
      }
    }
    corePluginRegistry.add("myCorePlugin", MyCorePlugin);

    const { alice, bob, network } = setupCollaborativeEnv();
    network.concurrent(() => {
      setCellContent(alice, "A1", "Hello");
      //@ts-ignore
      bob.dispatch("MY_CMD_1");
    });
    expect(numberCall).toEqual(1);
    uiPluginRegistry.remove("myUIPlugin");
    corePluginRegistry.remove("myCorePlugin");
  });
});
