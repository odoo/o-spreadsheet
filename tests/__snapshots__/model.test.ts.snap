// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model export formula with unbound zone stays unbound 1`] = `
Object {
  "borders": Object {},
  "entities": Object {},
  "formats": Object {},
  "revisionId": "START_REVISION",
  "sheets": Array [
    Object {
      "areGridLinesVisible": true,
      "cells": Object {
        "A1": Object {
          "content": "=SUM(A3:3)",
          "format": undefined,
          "style": undefined,
        },
        "A2": Object {
          "content": "=SUM(A3:A)",
          "format": undefined,
          "style": undefined,
        },
      },
      "colNumber": 26,
      "cols": Object {},
      "conditionalFormats": Array [],
      "figures": Array [],
      "filterTables": Array [],
      "id": "Sheet1",
      "isVisible": true,
      "merges": Array [],
      "name": "Sheet1",
      "rowNumber": 100,
      "rows": Object {},
    },
  ],
  "styles": Object {},
  "uniqueFigureIds": true,
  "version": 13,
}
`;
